RecoHGCal/TICL/plugins/HGCDoublet.h

```diff
@@ -0,0 +1,176 @@
+
+  void tagAsInnerNeighbor(unsigned int otherDoublet) { theInnerNeighbors_.push_back(otherDoublet); }
+
+  bool checkCompatibilityAndTag(std::vector<HGCDoublet> &allDoublets,
+                                const std::vector<int> &innerDoublets, float minCosTheta,
+                                float minCosPointing = 1., bool debug = false) {
+    int nDoublets = innerDoublets.size();
+    int constexpr VSIZE = 4;
+    int ok[VSIZE];
+    double xi[VSIZE];
+    double yi[VSIZE];
+    double zi[VSIZE];
+    auto xo = getOuterX();
+    auto yo = getOuterY();
+    auto zo = getOuterZ();
+    unsigned int doubletId = theDoubletId_;
+
+    auto loop = [&](int i, int vs) {
+      for (int j = 0; j < vs; ++j) {
+        auto otherDoubletId = innerDoublets[i + j];
+        auto &otherDoublet = allDoublets[otherDoubletId];
+        xi[j] = otherDoublet.getInnerX();
+        yi[j] = otherDoublet.getInnerY();
+        zi[j] = otherDoublet.getInnerZ();
+      }
+      for (int j = 0; j < vs; ++j) {
+        ok[j] = areAligned(xi[j], yi[j], zi[j], xo, yo, zo, minCosTheta, minCosPointing, debug);
+        if (debug) {
+          std::cout << "Are aligned for InnerDoubletId: " << i + j << " is " << ok[j] << std::endl;
+        }
+      }
+      for (int j = 0; j < vs; ++j) {
+        auto otherDoubletId = innerDoublets[i + j];
+        auto &otherDoublet = allDoublets[otherDoubletId];
+        if (ok[j]) {
+          otherDoublet.tagAsOuterNeighbor(doubletId);
+          allDoublets[doubletId].tagAsInnerNeighbor(otherDoubletId);
+        }
+      }
+    };
+    auto lim = VSIZE * (nDoublets / VSIZE);
+    for (int i = 0; i < lim; i += VSIZE) loop(i, VSIZE);
+    loop(lim, nDoublets - lim);
+
+    if (debug) {
+      std::cout << "Found " << theInnerNeighbors_.size() << " compatible doublets out of "
+                << nDoublets << " considered" << std::endl;
+    }
+    return theInnerNeighbors_.empty();
+  }
+
+  int areAligned(double xi, double yi, double zi, double xo, double yo, double zo,
+                 float minCosTheta, float minCosPointing, bool debug = false) {
+    auto dx1 = xo - xi;
+    auto dy1 = yo - yi;
+    auto dz1 = zo - zi;
+
+    auto dx2 = getInnerX() - xi;
+    auto dy2 = getInnerY() - yi;
+    auto dz2 = getInnerZ() - zi;
+
+    // inner product
+    auto dot = dx1 * dx2 + dy1 * dy2 + dz1 * dz2;
+    // magnitudes
+    auto mag1 = std::sqrt(dx1 * dx1 + dy1 * dy1 + dz1 * dz1);
+    auto mag2 = std::sqrt(dx2 * dx2 + dy2 * dy2 + dz2 * dz2);
+    // angle between the vectors
+    auto cosTheta = dot / (mag1 * mag2);
+    if (debug) {
+      std::cout << "dot: " << dot << " mag1: " << mag1 << " mag2: " << mag2
+                << " cosTheta: " << cosTheta << " isWithinLimits: " << (cosTheta > minCosTheta)
+                << std::endl;
+    }
+
+    // Now check the compatibility with the pointing origin.
+    // TODO(rovere): pass in also the origin, which is now fixed at (0,0,0)
```

Comment:
should this todo be implemented before the PR is merged?

Reply:
Having imagined a review of 30+ days maybe I should have started.
Other people are currently committed to doing this so, so no.